src/Cake.NuGet/Install/Extensions/NuGetPackageManagerExtensions.cs

```diff
@@ -0,0 +1,258 @@
+{
+    internal static class NuGetPackageManagerExtensions
+    {
+        public static async Task<IEnumerable<NuGetProjectAction>> GetInstallProjectActionsAsync(
+            this NuGetPackageManager packageManager,
+            NugetFolderProject nuGetProject,
+            PackageIdentity packageIdentity,
+            ResolutionContext resolutionContext,
+            INuGetProjectContext nuGetProjectContext,
+            IEnumerable<SourceRepository> sources,
+            CancellationToken token)
+        {
+            if (nuGetProject == null)
+            {
+                throw new ArgumentNullException(nameof(nuGetProject));
+            }
+
+            if (packageIdentity == null)
+            {
+                throw new ArgumentNullException(nameof(packageIdentity));
+            }
+
+            if (resolutionContext == null)
+            {
+                throw new ArgumentNullException(nameof(resolutionContext));
+            }
+
+            if (nuGetProjectContext == null)
+            {
+                throw new ArgumentNullException(nameof(nuGetProjectContext));
+            }
+
+            if (sources == null)
+            {
+                throw new ArgumentNullException(nameof(sources));
+            }
+
+            if (!sources.Any())
+            {
+                throw new ArgumentException(nameof(sources));
+            }
+
+            if (packageIdentity.Version == null)
+            {
+                throw new ArgumentNullException(nameof(packageIdentity));
+            }
+
+            var localSources = new HashSet<SourceRepository>(new SourceRepositoryComparer());
+            localSources.Add(packageManager.PackagesFolderSourceRepository);
+            localSources.AddRange(packageManager.GlobalPackageFolderRepositories);
+
+            if (resolutionContext.DependencyBehavior == DependencyBehavior.Ignore)
+            {
+                var logger = new ProjectContextLogger(nuGetProjectContext);
+
+                // First, check only local sources.
+                var sourceRepository = await GetSourceRepositoryAsync(packageIdentity, localSources, logger, token);
+
+                // If not found, check remote sources.
+                if (sourceRepository == null)
+                {
+                    sourceRepository = await GetSourceRepositoryAsync(packageIdentity, sources, logger, token);
+                }
+
+                // If still not found, we just throw an exception.
+                if (sourceRepository == null)
+                {
+                    throw new InvalidOperationException(
+                        $"Unable to find version '{packageIdentity.Version}' of package '{packageIdentity.Id}'.");
+                }
+
+                return new[] { NuGetProjectAction.CreateInstallProjectAction(packageIdentity, sourceRepository, nuGetProject) };
+            }
+
+            const bool downgradeAllowed = false;
+
+            // Step-1 : Get metadata resources using gatherer
```

Comment:
Could we break out step 1 into it's own method somehow? A lot is happening here and the method would probably benefit from being smaller.